### PR TITLE
fix(core): stop RegexFilterProcessor's redact strategy from discarding its matches

### LINKED ISSUE #20444: [BUG] RegexFilterProcessor's redact strategy throws away its matches
### Describe the Bug

`redact` finds its matches, drops them on the floor, then rewrites the text by running every rule again as a plain `String.replace`. Two things fall out of that.

**One, it leaks.**

`findMatches` (`regex-filter.ts:196`) collects `{ rule, match, index }` for every rule. But on `redact`, `handleMatches` (`:266`) has no branch for it, so it just returns. Then `redactText` (`:212`) re-runs every rule in order, each one working on whatever the previous one left behind.

In the `pii` preset, `phone` (`:93`) comes before `credit-card` (`:103`). A card number with no separators matches both, so `phone` eats the first ten digits and nothing looks at the rest:

```
"card 4111111111111111"  ->  "card [PHONE]111111"
```

Last six digits go out in the clear. Same thing can happen with any two rules whose patterns can overlap, like `url` / `email` / `ip-address`.

**Two, nothing knows what got rewritten.**

That `matches` array in `processInput` (`:308`), `processOutputStream` (`:327`) and `processOutputResult` (`:347`) is only used to check "did anything match at all". There are three strategies and not one of them both redacts and reports.

- `warn` leaves the text alone. It logs a single line, `[RegexFilterProcessor] Matched patterns: email, phone`, which is a string for a human to read. No position, no length, nothing a program can consume.
- `block` aborts the request. The error it throws does carry the match list, but the matched text itself is replaced with `'[REDACTED_MATCH]'` (`:258`), and more to the point the conversation just fails. Nothing got redacted and carried on.
- `redact` is the only strategy that redacts and lets the conversation continue. And it has no way to tell you what it changed. Not in the return value, not through a callback, not in a trace, not in a log.

What makes this look like an oversight rather than a decision: `Processor.onViolation` already exists for exactly this (`processors/index.ts:580`), described as "invoked when this processor detects a violation, regardless of strategy. Use for side effects like alerting, logging to external systems". `RegexFilterProcessor` never sets or calls it. `block` reaches it only because `runner.ts:56` invokes it when it catches a TripWire, and `warn` and `redact` don't report through it at all. `CostGuardProcessor` does call it itself on its non-block path (`cost-guard.ts:284`), so there's already a pattern to follow.

So if you're using `redact` as a privacy control, you have no record of what it removed.

**Why I'm filing these as one thing**

Same root cause: finding and rewriting are two separate passes. That's also why you can't just bolt reporting on top. Take `card 4111111111111111` — `findMatches` says it found both `phone` and `credit-card`, but `redactText` only ever applied `phone`. Log that and you've written "card number redacted" while six digits walked out the door. The two have to agree before reporting means anything.

### Steps To Reproduce

```ts
import { RegexFilterProcessor } from '@mastra/core/processors';

const filter = new RegexFilterProcessor({ presets: ['pii'], strategy: 'redact' });

const messages = [
  {
    id: 'm1',
    role: 'user' as const,
    content: { format: 2, parts: [{ type: 'text' as const, text: 'card 4111111111111111' }] },
    createdAt: new Date(),
  },
];

const result = filter.processInput({ messages, state: {}, retryCount: 0 } as any) as any;

console.log(result[0].content.parts[0].text);
// "card [PHONE]111111"   <- last six digits still there
// want: "card [CREDIT_CARD]"
```

A few variants:

| Input | What you get today |
| - | - |
| `card 4111111111111111` | `card [PHONE]111111` |
| `pay 4111111111111111 now` | `pay [PHONE]111111 now` |
| `card 4111-1111-1111-1111` | `card [CREDIT_CARD]` — fine, the dashes stop `phone` from matching |

That's the whole reproduction. One processor, one message, no agent, model, or network involved, and it runs against the published `@mastra/core@1.55.0`. Happy to push it as a repo if you'd rather have one.

For the second problem there's nothing to reproduce against. Try to hook up any kind of observer and you'll find there's no hook, no emitted chunk, no span, no log line to attach it to.

### Expected Behavior

Two things:

1. Overlapping matches get redacted as one span, with nothing left over. `card 4111111111111111` should come out as `card [CREDIT_CARD]`.
2. `redact` reports what it rewrote through `onViolation`, the same callback `block` already reaches through the runner. Rule name, offset and length per span is plenty. Leave the matched values out by default though — `block` already hides them from `TripWire` metadata (`:258`), and an audit log that copies the data you're trying to protect kind of defeats the point. An opt-in flag can add them for callers whose sink is as protected as the original.

### Additional Context

- #20234 asked for deterministic PII redaction without an LLM, and already mentions that "the `redact` strategy doesn't report what it matched". That one proposed a separate processor, which the team passed on in #20236 in favour of improving `RegexFilterProcessor`. So this issue is the same ask narrowed down to the two actual defects in the processor we already have. Leaving #20234 open for the bigger proposal.
- #20234 also argued that detection events should flow through the existing `Processor.onViolation` / `ProcessorViolation` mechanism rather than a new per-processor callback. Same position here, which is why the second point above asks for `onViolation` rather than a new option.
- Both reproduce in the package's own tests. `regex-filter.test.ts` has no overlapping-pattern coverage right now.
- Unrelated but nearby: `PIIDetector.applyRedactionMethod` (`pii-detector.ts:499`) sorts detections in reverse and splices, which also assumes the spans don't overlap. Probably worth its own look — detections there come from an LLM too, so it's not the same fix.

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have included sufficient information for the team to reproduce and understand the issue

### PR DESCRIPTION
**Pull request by jee14**
## Description

`RegexFilterProcessor`'s `redact` strategy found its matches, dropped them, and rewrote the text by running every rule again as a plain `String.replace`. Two problems came out of that one cause, so they're fixed together in two commits.

**Commit 1 — overlapping matches no longer leak** (`fix`, patch)

Rules ran one at a time, each over whatever the previous one left behind. In the `pii` preset `phone` is declared before `credit-card`, so a card number with no separators lost only its first ten digits:

```ts
new RegexFilterProcessor({ presets: ['pii'], strategy: 'redact' });

// before: "card 4111111111111111"  ->  "card [PHONE]111111"
// after:  "card 4111111111111111"  ->  "card [CREDIT_CARD]"
```

Matches are now collapsed into disjoint regions, unioning any that overlap, and each region is replaced once using the longest match's replacement.

**Commit 2 — `redact` reports what it rewrote** (`feat`, minor)

Reporting goes through the existing `Processor.onViolation`, the same callback `block` already reaches when the runner catches its `TripWire`, and the one `CostGuardProcessor` calls itself on its non-block path.

```ts
import { RegexFilterProcessor, type RegexRedactionDetail } from '@mastra/core/processors';

const filter = new RegexFilterProcessor({ presets: ['pii'], strategy: 'redact' });

filter.onViolation = async ({ detail }) => {
  const redaction = detail as RegexRedactionDetail;

  for (const entry of redaction.redactions) {
    await auditLog.write({
      phase: redaction.phase, // processInput | processOutputStream | processOutputResult
      messageId: redaction.messageId,
      rule: entry.rule, // 'credit-card'
      offset: entry.index,
      length: entry.length,
    });
  }
};
```

One report per redacted message, message part, or stream chunk, so offsets always refer to a single piece of text. When several rules matched the same span, the winning rule is `rule` and every rule that matched is listed in `overlappingRules`.

Async callbacks are awaited. With no callback attached the `redact` path stays synchronous, so nothing changes for existing callers.

**Values are withheld by default.** Reports carry offsets and rule names, not the matched text. An audit trail that copies the data it protects widens the exposure it was added to narrow, which is why `block` already replaces `match` with `'[REDACTED_MATCH]'` in its `TripWire` metadata. `includeRedactedValues: true` opts in for callers whose sink is as protected as the original.

### Two smaller behaviour changes to redaction

- A replacement referencing capture groups (`$1`, `$&`) falls back to the replacement string as written when the rule cannot match the text it matched in isolation, which happens with a lookbehind or lookahead. The region is still redacted either way — the previous code silently left such a region untouched under the new span logic, so this path is explicit and covered by a test.
- A rule that only matches empty strings no longer inserts its replacement between every character.

### Two things worth calling out

- **`warn` still only logs.** This PR wires `redact`. `CostGuardProcessor` does report through `onViolation` on its `warn` path, so wiring `warn` the same way would be consistent — happy to do it here or as a follow-up, whichever you prefer.
- **The `redact` path still scans the text twice**, once to decide whether anything matched and once inside the rewrite. That predates this PR and is unchanged; consolidating it seemed like separate work.

### Verification

- `pnpm --filter ./packages/core vitest run src/processors` — 39 files, 1042 passed, 1 skipped
- `pnpm --filter ./packages/core check` and `oxlint` / `eslint` on the changed files — clean
- `pnpm validate` in `docs` — 585 files pass. `lint:prose` could not run locally: `scripts/vale/bin/vale` fails with `mdx2vast not found` on untouched files too.

Two checks on the span rewrite specifically, since it changes a code path that had no overlap coverage:

- **Behaviour preserved.** Every string literal in `regex-filter.test.ts` (107 of them) produces byte-identical output under the old sequential `replace` and the new span logic. No existing literal contained overlapping matches, so the only behaviour that changes is the buggy case.
- **Tests actually catch the bug.** 13 of the 21 tests added in commit 1 fail against the pre-fix implementation. The other 8 are regression guards that pass both before and after (separated card numbers, capture groups, non-global patterns, default replacement, touching-but-not-overlapping spans).

Also corrects the agent example on the reference page, which used a `processors: { input: [...] }` option that does not exist — `Agent` takes `inputProcessors` and `outputProcessors`. `cost-guard-processor.mdx` has the same mistake but is left alone as out of scope.

## Related issue(s)

Fixes #20444

Context: #20234 asked for deterministic, LLM-free PII redaction and #20236 proposed a separate processor, which the team declined in favour of improving `RegexFilterProcessor`. This is that, narrowed to the two defects in the existing processor. #20234 stays open for the broader proposal.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [x] I have linked the related issue(s) in the description above
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have addressed all Coderabbit comments on this PR

## ELI5
If two regex “redact” rules overlap, the old code could redact only part of the sensitive text—leaving other pieces visible. Now it redacts the whole overlapping region in one go and also tells you exactly what was removed and where.

## What changed
- Merges overlapping regex matches into disjoint spans and applies redaction once per span, using the replacement from the longest match (prevents partial redaction/leakage).
- Fixes `$` capture-group replacement resolution: only expands captures when the matched slice can be fully and safely re-covered; otherwise falls back to a full-region replacement strategy.
- Prevents rules that match only empty strings from inserting replacements between characters.
- Reports `redact` activity via `Processor.onViolation`, including:
  - processor phase and message/unit context (part vs chunk/stream),
  - message ID,
  - rule name(s),
  - offsets/lengths for the affected region,
  - replacement details,
  - `overlappingRules` when multiple rules overlap the same span.
- Excludes matched/redacted values from violation details by default; adds `includeRedactedValues` option (opt-in) to include the value in `RegexRedactionDetail`.
- Awaits async `onViolation` handlers, while keeping behavior synchronous when no callback is provided.
- Adds/updates documentation, types, and tests for overlapping redaction, streaming/structured reporting, and replacement-resolution edge cases.
- Updates an agent documentation example to use `inputProcessors`/`outputProcessors`.
- Re-exports `RegexRedaction` and `RegexRedactionDetail` types and extends `RegexFilterOptions` with `includeRedactedValues`.

### CODE REVIEW

File: packages/core/src/processors/processors/regex-filter.ts [hunk 1]
```diff
@@ -209,13 +316,129 @@ export class RegexFilterProcessor implements Processor<'regex-filter', RegexFilt
...
+  private resolveReplacement(text: string, region: RedactionRegion): string {
+    const { rule } = region.owner;
+    const replacement = rule.replacement ?? '[REDACTED]';
+    if (region.matches.length > 1 || !replacement.includes('$')) return replacement;
+
+    const matched = text.slice(region.start, region.end);
+    if (!compilePattern(rule).test(matched)) return replacement;
+
+    return matched.replace(compilePattern(rule), replacement);
+  }
```
[the remark this answers is not in the document]
Comment (jee14) [reply packages/core/src/processors/processors/regex-filter.ts:383]: Good catch, this was reachable. Fixed in effe4e2410.

A rule whose match length depends on trailing context matches a shorter span in isolation, and `test()` still returns true:

```
rule    /\d+(?=\d)/g  ->  '<$&>'
input   "1234"
region  "123"        (greedy \d+ backtracks so the lookahead has a digit left)
before  "<12>34"     the 3 survives
after   "<$&>4"      region replaced whole
```

Took the suggested guard as-is. Added a regression test (`falls back when the pattern covers less of the region on its own`) that asserts the output and that no part of the region leaks.
